pkg/devconsole/proxy/ssrf_utils.go

```diff
@@ -0,0 +1,142 @@
+package proxy
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// List of allowed URLs for SSRF protection
+var allowedURLs = []string{
+	// ArtifactHub Public API
+	"https://artifacthub.io/api/v1",
+
+	// Tekton Results installed in the cluster
+	"http://tekton-results-api-service",
+	"https://tekton-results-api-service",
+}
+
+func isAllowedURL(requestURL string) bool {
+	for _, allowedURL := range allowedURLs {
+		if strings.HasPrefix(requestURL, allowedURL) {
+			return true
+		}
+	}
+	return false
+}
+
+// Helper function to check if an IP is private
+func isPrivateIP(ip net.IP) bool {
+	privateIPBlocks := []*net.IPNet{
+		// IPv4 private ranges (RFC 1918)
+		{IP: net.IPv4(10, 0, 0, 0), Mask: net.CIDRMask(8, 32)},
+		{IP: net.IPv4(172, 16, 0, 0), Mask: net.CIDRMask(12, 32)},
+		{IP: net.IPv4(192, 168, 0, 0), Mask: net.CIDRMask(16, 32)},
+		// IPv4 link-local addresses (169.254.0.0/16)
+		{IP: net.IPv4(169, 254, 0, 0), Mask: net.CIDRMask(16, 32)},
+		// IPv6 link-local addresses (fe80::/10)
+		{IP: net.ParseIP("fe80::"), Mask: net.CIDRMask(10, 128)},
+		// IPv6 private range (fc00::/7)
+		{IP: net.ParseIP("fc00::"), Mask: net.CIDRMask(7, 128)},
+		// IPv6 loopback address (::1)
+		{IP: net.ParseIP("::1"), Mask: net.CIDRMask(128, 128)},
+	}
+
+	for _, block := range privateIPBlocks {
+		if block.Contains(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+// Whitelist-based check to only allow public IP addresses (e.g., from the internet)
+func isPublicIP(ip net.IP) bool {
+	if ip.IsLoopback() || ip.IsLinkLocalUnicast() || ip.IsPrivate() || isPrivateIP(ip) {
+		return false
+	}
+	return true
+}
+
+// Helper function to resolve the DNS hostname and check consistency
+func resolveHostname(hostname string) ([]net.IP, error) {
+	ips, err := net.LookupIP(hostname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve hostname: %v", err)
+	}
+
+	// Introduce DNS rebinding protection: re-check the IPs after a small delay to prevent DNS rebinding
```

Comment:
If this is being resolved by a pod, using the default resolver, we force a min ttl of 30 seconds anyway, so this won't ever matter.

But it can't hurt (other than adding a 2s delay) in case that behavior changes in the future.